[쇼핑 장바구니 3, 4단계] 뭉치 미션 제출합니다.

m6z1 opened the pull request:
# 셀프 체크리스트

 <!--
 모든 항목을 점검한 후 리뷰를 요청하면, 보다 의미 있는 피드백을 받을 수 있습니다.
 -->

- [x] 프로그램이 정상적으로 작동하는가?
- [x] 모든 테스트가 통과하는가?
- [ ] 이전에 받은 피드백을 모두 반영하였는가?
- [x] 코딩 스타일 가이드를 준수하였는가?
    - [x] IDE 코드 자동 정렬을 적용하였는가?
    - [x] 린트 검사를 통과하였는가?
- [x] 프로그래밍 요구 사항을 준수하였는가?
- [x] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

# 어떤 부분에 집중하여 리뷰해야 할까요?

 <!--
 학습 목표에 맞게 구현하였는지 점검하고, 피드백을 받을 준비가 되었는지 확인하세요.
 정답을 묻기보다, 자신의 생각과 의사 결정 과정을 공유하며 피드백을 통해 논의하는 것이 좋습니다.
 -->
 
안녕하세요 페로로
날씨가 좋은 월요일(?)입니다 ,,,
2단계를 빠르게 머지해주셨는데 저의 능력 부족으로 인해서 3, 4단계 리뷰 요청을 마감 시간 전에 보내게 됐습니다 ㅎㅎ......

장바구니 <-> 상품 목록 / 상품 목록 <-> 상품 상세 간 이동이 생길 때마다 데이터를 모두 호출하는 것은 불필요하다고 생각했고, 성능이 떨어질 것 같아 이 부분을 중점적으로 개선하기 위해 노력했습니다.
하지만, 작성하다보니 뷰모델이 굉장히 더러워졌습니다. ~(과연 성능을 위해서 인지,, 저의 코드가 더러운지는 ,,, ㅎㅎ)~

미션 요구 사항을 모두 충족하기 위해서 열심히 노력했으며 상태 관리 중점으로 피드백 받고 싶습니다..!
이번 미션 또한 피드백을 받은 후 빠진 테스트 및 리팩토링을 진행하겠습니다.

이번 리뷰도 잘 부탁드립니다 ,,, ㅠㅠㅠㅠㅠ
~(한번만 더..!! 절 리뷰이로 받아들여주세요!!!)~
![image](https://github.com/user-attachments/assets/7fdb45fe-4b58-441b-8a16-9bd49b5936b5)


## 스크린샷

https://github.com/user-attachments/assets/25d1e97c-84e9-4f35-bdde-1ec5ed77a16a

m6z1 committed:
docs: 기능 요구 사항 정리

m6z1 committed:
refactor: result 콜백 매개변수 네이밍 수정

m6z1 committed:
refactor: 상품 목록 조회 호출 로직 수정

m6z1 committed:
chore: activity exported 속성 수정

m6z1 committed:
refactor: 상품 목록 아이템 추가 함수 네이밍 수정

m6z1 committed:
refactor: 중복되는 값 변수에 할당

- ProductsViewModel에서 중복되는 라이브 데이터 값 할당 후 사용

m6z1 committed:
refactor: 함수 네이밍 명시적으로 수정

m6z1 committed:
feat: 각 상품에 + 버튼 ui 구현

m6z1 committed:
feat: 상품 목록 뷰 장바구니 로직 추가

- 각 상품에서 + - 버튼을 통해 수량을 조절할 수 있다.
- 장바구니에 담근 총 수량이 보인다.

m6z1 committed:
feat: 상품 수량 공용 컴포넌트 Ui 구현

m6z1 committed:
feat: 상품 상세 화면 수량 조절 버튼 구현

m6z1 committed:
feat: 상품 상세 화면 수량만큼 선택한 총 가격 출력 구현

m6z1 committed:
refactor: 장바구니 상품 조회 반환 타입 수정

m6z1 committed:
feat: 장바구니 화면 수량 별 가격 출력 구현

m6z1 committed:
feat: 장바구니 상품 제거 / 장바구니 수량 빼는 로직 분리

m6z1 committed:
feat: 장바구니에 담은 수량 조절 로직 구현

m6z1 committed:
feat: 상품 상세 화면에서 뒤로 갔을 때 상태 일치 로직 구현

m6z1 committed:
feat: 장바구니 화면에서 뒤로 갔을 때 상태 일치 로직 구현

m6z1 committed:
build: room 라이브러리 추가

m6z1 committed:
docs: 기능 요구 사항 정리

m6z1 committed:
refactor: application class 생성

m6z1 committed:
build: Room 라이브러리 버전 2.7.1 업데이트

m6z1 committed:
feat: 앱 재시작에도 장바구니 데이터 유지

m6z1 committed:
feat: 앱 재시작에도 상품 목록 장바구니 개수 유지

m6z1 committed:
chore: 불필요한 파일 제거

m6z1 committed:
feat: 최근 본 상품 로직 구현

m6z1 committed:
refactor: RecyclerView 아이템 데코레이터 구현

m6z1 committed:
feat: 앱 재시작돼도 최근 본 상품 목록 유지

m6z1 committed:
refactor: string 리소스 추출

m6z1 committed:
feat: 상품 상세화면에서 마지막으로 본 상품 출력 로직 구현

m6z1 committed:
feat: 마지막으로 본 상품일 경우 마지막으로 본 상품 ui 가시성 처리

m6z1 committed:
fix: 마지막에 본 상품 클릭 시 마지막에 본 상품 ui 출력되는 이슈 해결

m6z1 commented:
@coderabbitai pause

m6z1 committed:
feat: 마지막으로 본 상품 페이지 뒤로가기 시 상품 목록 이동 구현

m6z1 committed:
fix: 최근 본 상품 이전의 상품 개수 업데이트 안 되는 이슈 해결

m6z1 committed:
chore: 패키지 분리

m6z1 committed:
build: 네트워크용 라이브러리 추가

m6z1 committed:
chore: 패키지 이동

m6z1 committed:
feat: 상품 목록 값 서버 통신으로 호출

m6z1 committed:
refactor: 불필요한 함수 제거

m6z1 commented:
@coderabbitai resume

m6z1 committed:
refactor: 상품 목록 뷰 수량 조절 컴포넌트 사용

m6z1 committed:
refactor: 함수 분리

m6z1 committed:
test: 변경된 프로덕션 코드에 테스트 성공하도록 수정

m6z1 committed:
chore: 사용하지 않는 코드 제거

m6z1 committed:
test: mock 상품 서버 테스트 추가

m6z1 committed:
refactor: mock server 클래스 네이밍 수정

laco-dev commented on app/src/main/java/woowacourse/shopping/data/product/local/database/ProductDatabase.kt:
```diff
@@ -0,0 +1,15 @@
+    entities = [ProductEntity::class, RecentWatchingEntity::class],
+    version = 2,
+)
+abstract class ProductDatabase : RoomDatabase() {
+    abstract fun recentWatchingDao(): RecentWatchingDao
+}
```
데이터베이스 객체를 두개로 만든 이유가 있을까요?
이렇게 하면 다른 도메인이 생길 때 마다 무수히 많은 보일러 플레이트를 만들어야 하지 않을까 싶어요.

m6z1 replied:
서로 성격이 다르다 생각해 db 객체를 두 개로 만들었는데, 앱 내에서는 db가 하나로 엔티티를 여러 개 사용하면 될 것 같다고 판단해 디비를 하나로 사용하도록 수정했습니다!

m6z1 replied:
성격이 다른 디비를 여러 개 만들어야 한다 생각했는데, 하나의 디비에 대해 엔티티를 여러 개 둘 수 있다는 점을 간과했었습니다 😓

앱에서 디비 하나로 합쳐 사용할 수 있도록 수정했습니다!

ada7aa9dbccfab20a1d66fc9166d12ac14c1452b

laco-dev commented on app/src/main/java/woowacourse/shopping/data/product/local/entity/RecentWatchingEntity.kt:
```diff
@@ -0,0 +1,13 @@
+import androidx.room.PrimaryKey
+import woowacourse.shopping.domain.product.Product
+
+@Entity(tableName = "recent_watching")
+data class RecentWatchingEntity(
+    @PrimaryKey val productId: Long,
+    @Embedded val product: Product,
+    val watchedAt: Long = System.currentTimeMillis(),
+)
```
최근 본 상품 테이블에 상품을 데이터베이스에 그대로 넣을 때 고민해야 할 내용이 있어요.
상품과 최근 본 상품 테이블의 데이터가 서로 불일치 하는 문제가 발생할 수 있는데요.

가령 동일한 상품인데 한쪽은 5,000원 한쪽은 10,000원일 수 있죠.
이 문제는 어떻게 해결하는게 좋다고 생각하시나요?

m6z1 replied:
최근 본 상품엔 상품 id 만 넘기고 해당 최근 본 상품을 사용할 때는 상품 id 를 통해 데이터를 조회할 수 있도록 한다면 상태 불일치 문제를 해결하면서 단일 출처를 유지할 수 있을 것 같습니다!

상품 id 를 통해 상품 정보를 조회할 수 있도록 수정했습니다!
0bce4322abe77cd7afe520f1ad3024f9c5479c54

laco-dev commented on app/src/main/java/woowacourse/shopping/data/product/remote/MockProductServer.kt:
```diff
@@ -0,0 +1,56 @@
+    val mockWebServer = MockWebServer()
+
+    fun start(port: Int) {
+        thread {
+            mockWebServer.start(port)
+        }
```
초기화 하는데 스레드가 필요한 이유가 무엇인가요?

m6z1 replied:
`android.os.NetworkOnMainThreadException` 에러가 발생해, 네트워크 작업은 메인 스레드에서 수행할 수 없다는 점을 확인했습니다.

기존 코드는 MockProductServer의 init 블록에서 서버를 초기화했지만,

네트워크 요청을 수행하는 시점은 서버 자체를 구동하는 목서버가 아닌, 실제로 데이터를 가져와야 하는 remote 계층의 책임이란 생각이 들어 서버 호출이 필요한 remoteStorage 에서 스레드를 사용하도록 수정했습니다!

관심사를 분리함으로써 mock 서버는 서버 역할에 집중하고, 실제 네트워크 요청은 storage 계층이 맡아 보다 명확한 구조가 되도록 개선했습니다.

208ae0218c34db4c934d69f570cefddc1caa1cd3

laco-dev commented on app/src/main/java/woowacourse/shopping/data/product/storage/ProductRemoteStorage.kt:
```diff
@@ -0,0 +1,41 @@
+        mockProductServer.start(12345)
+    }
+
+    override fun load(
+        lastProductId: Long?,
+        size: Int,
+    ): List<ProductEntity> {
```
Storage의 역할은 무엇인가요?
Repository와는 어떤 차이가 있는 것인지 잘 이해가 안되서요.

Remote에 접근하기 위한 인터페이스라면 local.xxx을 참조하면 안됩니다.

m6z1 replied:
데이터를 가져오는 근원체라 생각했습니다.
로컬에서 값을 가져올 수도 있고 remote 에서 가져올 수도 있고 이를 분리하기 위한 수단으로 storage 라는 객체를 생성했습니다.

하지만, 말씀해주신 것처럼 remote 에 접근하는데 local 의 엔티티를 참조하고 있어 remote 에서 사용할 응답 dto 를 생성해 local 을 참조하지 않도록 수정했습니다!

a8a0d6bc28ae07754d8c5e0de1d68faef385ba83

laco-dev commented on app/src/main/java/woowacourse/shopping/view/common/ResultFrom.kt:
```diff
@@ -0,0 +1,10 @@
+package woowacourse.shopping.view.common
+
+enum class ResultFrom {
+    SHOPPING_CART_BACK,
+    PRODUCT_DETAIL_BACK,
+    PRODUCT_RECENT_WATCHING_CLICK,
+    ;
+
+    val RESULT_OK: Int = ordinal + 1
+}
```
화면간의 이동 관계를 common으로 모으는 것은 좋지 않습니다.
가령 화면이 100개, 200개가 넘어가면 답도 없이 불어나는 코드를 볼 수 있고 비즈니스 로직에 따라 다르게 동작하는 등 복잡도가 많이 올라갑니다.

m6z1 replied:
각 화면에서 이동에 대한 Result 상수 값을 선언해 사용하도록 수정했습니다!

75fa1e5ba7e02e7c11a8f316468a9988e59f193a

laco-dev commented on app/src/main/res/layout/component_product_quantity.xml:
```diff
@@ -0,0 +1,78 @@
+
+        <TextView
+            android:id="@+id/productPlusQuantityButton"
+            android:layout_width="0dp"
+            android:layout_height="42dp"
```
왼쪽 오른쪽 버튼이 완전 동일하기에 하나를 바꿀 때 다른 하나를 바꾸지 못하는 실수가 발생할 수 있어요.
style을 찾아보고 한번 적용해보세요 :)
```xml
android:style="xxx"

m6z1 replied:
styles 에 버튼 스타일을 지정해 사용했습니다! : ) 

640a3c65ca571cee2d083aecd182a1984d571065

laco-dev commented on app/src/main/res/layout/item_shopping_cart_product.xml:
```diff
@@ -55,18 +55,33 @@
             app:layout_constraintTop_toBottomOf="@id/shoppingCartProductName"
             tools:src="@drawable/product_image_example" />
 
+        <androidx.constraintlayout.widget.ConstraintLayout
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:layout_marginBottom="6dp"
+            app:layout_constraintBottom_toTopOf="@id/shoppingCartPrice"
+            app:layout_constraintEnd_toEndOf="@id/shoppingCartPrice">
+
+            <include
+                android:id="@+id/shoppingCartQuantityComponent"
+                layout="@layout/component_product_quantity" />
+
+        </androidx.constraintlayout.widget.ConstraintLayout>
```
ConstraintLayout로 한번 감싸서 처리한 이유가 무엇일까요?

m6z1 replied:
제약 조건을 거는데 xml 디자인뷰에서 반영이 안 되는 것처럼 보여 레이아웃을 한 번 감쌌었습니다.
하지만, 레이아웃 없이도 제약조건이 잘 반영되는 것을 확인해 모든 include 에 대해 감싼 layout을 제거했습니다!

7e4aff9a0a4dce9613107e21af44cc2bb871e54b

laco-dev commented on app/src/test/java/woowacourse/shopping/data/MockProductServerTest.kt:
```diff
@@ -0,0 +1,83 @@
+
+    private fun String.allSpaceTrim(): String = replace("\\s".toRegex(), "")
+
+    @AfterEach
+    fun tearDown() {
+        mockProductServer.shutDown()
+    }
```
테스트마다 해제하는 방법이 서로 다른 것 같네요?

m6z1 replied:
해당 mock 서버 테스트에서는 mockk() 객체를 사용하지 않아 clearAllMocks() 호출은 생략하고, 대신 실제 서버를 해제하기 위해 mockProductServer.shutDown()을 호출했습니다..!
테스트마다 사용하는 mock 방식이 다르기 때문에 해제 방식도 달라졌는데, 혹시 리뷰 주신 내용은 해제 방식의 통일성을 맞춰야 한다는 의미일까요?

혹은 `clearAllMocks()` 를 이 함수에서도 동일하게 호출해달라는 말씀이신 것일지 궁금합니다 😭

laco-dev commented on app/src/test/java/woowacourse/shopping/data/MockProductServerTest.kt:
```diff
@@ -0,0 +1,83 @@
+package woowacourse.shopping.data
+
+import okhttp3.OkHttpClient
+import okhttp3.Request
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.jupiter.api.AfterEach
+import org.junit.jupiter.api.Test
+import woowacourse.shopping.data.product.remote.MockProductServer
```
테스트 코드와 구현 코드의 패키지를 동일하게 맞춰주세요.

m6z1 replied:
해당 내용 반영했습니다!

426c6a73169226129b4e579df84c2b655d42b52b

laco-dev commented on app/src/test/java/woowacourse/shopping/view/product/ProductsViewModelTest.kt:
```diff
@@ -57,6 +80,35 @@ class ProductsViewModelTest {
+            lambda<(Result<List<Product>>) -> Unit>().invoke(Result.success(RECENT_PRODUCTS))
+        }
+
+        every {
+            shoppingCartRepository.fetchSelectedQuantity(any<List<Product>>(), captureLambda())
+        } answers {
+            lambda<(Result<List<ShoppingCartProduct>>) -> Unit>().invoke(Result.success(emptyList()))
+        }
```
이 패턴이 계속해서 반복된다면 맞춤 함수를 만들어서 가독성을 올려볼 수도 있겠어요! 

```kotlin
every { ... } returns(...)
```

laco-dev commented on app/src/main/java/woowacourse/shopping/view/productDetail/ProductDetailClickListener.kt:
```diff
@@ -1,7 +1,12 @@
+
+    fun onAddingToShoppingCart()
+
+    fun onRecentProduct(product: Product)
```
인터페이스로 Click이라는 점이 드러내는 것과 함수 이름으로 Click으로 드러내는 것 무엇이 좋을까요?

가령 액티비티가 아래 2개의 리스너를 구현하고 있다고 가정해보세요.
```kotlin
interface ProductDetailClickListener {
  fun onCloseButton()
}

interface ProductDetailLongClickListener {
  fun onCloseButton()
}
```

m6z1 replied:
인터페이스로 click 이라는 점이 드러날 때의 장점으론 클릭을 모아둔 인터페이스인 것을 알 수 있을 것 같습니다.

하지만, 아래 알려주신 예시 코드와 같이 인터페이스에서 세부적인(길고 짧음 등)네임까지 알게 된다면,
불필요하게 인터페이스가 여럿 생기며, 함수 네이밍이 중복될 수 있는 우려가 있을 것 같습니다.
이는 곧 유지보수성과 확장성이 좋지 않을 것 같습니다.

함수 이름에서 click 이란 것을 드러낸다면 인터페이스는 이벤트 리스너를 모아둘 수 있는 관심사가 될 것 같습니다. 또한, 함수에서 정확히 어떤 식의 동작(길게 누르는 닫기 버튼 / 짧게 누르는 닫기 버튼) 등을 나타낼 수 있을 것 같습니다. 함수에 드러낸다면 더 명확히 구체적으로 명시하여 가독성이 좋아질 것 같습니다!

따라서, 인터페이스에서 구체적으로 어떤 이벤트인지 명시했던 것을 제거하고 함수에서 드러나도록 수정했습니다.

a19cc2c4107c00cb1d503c9e808a4dcfa49cf34b

laco-dev commented on app/src/main/java/woowacourse/shopping/view/productDetail/ProductDetailActivity.kt:
```diff
@@ -35,64 +35,94 @@ class ProductDetailActivity :
+        val isLastWatching: Boolean =
+            intent.getSerializableExtraData(EXTRA_IS_LAST_WATCHING) ?: false
+        viewModel.updateProduct(product, isLastWatching)
+        bindViewModel(product)
```
이 화면을 진입하기 위해 Product라는게 반드시 필요하다면 이것은 논리적인 오류로 개발자가 반드시 잡을 수 있도록 오류를 발생시키는 것도 방법입니다.

그렇게 하면 코드 처리가 더 간결해 질 수 있다는 뜻인데요.
이후 상품 정보는 뷰모델의 팩토리를 이용해 생성자로 넣어줄 수 있습니다.

지금 구조에서는 화면이 회전할 때 마다 updateProduct가 반복해서 호출이 되는 문제가 발생합니다.

m6z1 replied:
뷰모델 생성자 매개변수로 product를 추가해 무조건 product 가 없으면 에러가 뜨도록 수정했습니다!
> 지금 구조에서는 화면이 회전할 때 마다 updateProduct가 반복해서 호출이 되는 문제가 발생합니다.

해당 내용 저번 리뷰 때 지적해주셨는데 이번에도 놓쳤네요 😭 장치 구성 변경 고려해서 로직 짤 수 있도록 항상 명심하겠습니다!

fa7f9f2643cccc452cb9da8dfd68950568e8ef16

laco-dev commented on app/src/main/java/woowacourse/shopping/view/productDetail/ProductDetailActivity.kt:
```diff
@@ -35,64 +35,94 @@ class ProductDetailActivity :
         binding.lifecycleOwner = this
         binding.productDetailEventListener = this
+        binding.productDetailQuantityComponent.product = product
+        binding.productDetailQuantityComponent.productQuantityClickListener = this
```
vm.xxx을 이용해 바인딩을 연결해 줘도 되지 않을까요?

m6z1 replied:
해당 내용 반영했습니다!

7e4aff9a0a4dce9613107e21af44cc2bb871e54b

laco-dev commented on app/src/main/java/woowacourse/shopping/view/productDetail/ProductDetailActivity.kt:
```diff
@@ -35,64 +35,94 @@ class ProductDetailActivity :
+        }
+
+        viewModel.quantity.observe(this) { quantity: Int ->
+            binding.productDetailQuantityComponent.quantity = quantity
```
이 값들도 vm.xxx을 활용해보면 될 것 같아요.

m6z1 replied:
수량 변경 컴포넌트를 활용하는 모든 곳을 데이터바인딩으로 수정했습니다!

7e4aff9a0a4dce9613107e21af44cc2bb871e54b

laco-dev commented on app/src/main/java/woowacourse/shopping/view/productDetail/ProductDetailActivity.kt:
```diff
@@ -35,64 +35,94 @@ class ProductDetailActivity :
+            Intent().apply {
+                putExtra("updateProduct", viewModel.product.value)
+            }
+        setResult(ResultFrom.PRODUCT_DETAIL_BACK.RESULT_OK, intent)
```
> 화면을 종료하라

라는 액션은 ProductDetailEvent로 봐야할까요? 아니면 지금처럼 여기서 바로 로직을 처리해도 될까요?
만약 필요한 데이터가 뷰모델에 private으로 선언되어 있었다면 이를 위해 public으로 열어야 하는 것일까요?
그렇지 않다면 이정도는 귀찮으니 그냥 넘어가도 되는지?

m6z1 replied:
상황에 따라 다를 것 같습니다.

액션에 대해 비즈니스 로직이나 상태 변화가 필요할 경우엔 이벤트로 통해 viewModel 을 거쳐 뷰가 변경되어야 할 것 같습니다.

예를 들어 addProductToShoppingCart 같은 뷰모델 메서드를 호출해 상태를 변경해야 한다면 이벤트를 뷰모델에 전달해야 한다고 생각합니다. 하지만, 데이터 바인딩을 통해 이미 뷰모델의 값을 구독하고 있고, 해당 값을 바로 넘길 수 있는 환경이 된다면 뷰에서 바로 처리해도 될 것 같습니다. 

기존 코드에서는 값을 옵저빙하지 않고 바로 꺼내 쓰고 있어 문제가 될 것 같습니다.

> 만약 필요한 데이터가 뷰모델에 private으로 선언되어 있었다면 이를 위해 public으로 열어야 하는 것일까요?

이렇게 된다면 테스트 때문에 프로덕션 코드가 바뀌듯이 뷰모델이 가지고 있는 상태를 직접 꺼내쓰는 것이 될 것 같아 캡슐화가 깨질 것 같습니다. -> 간접적으로 옵저빙을 통해 관찰하는 것과 직접 멱살 잡아 꺼내 쓰는 것은 다르지 않을까.. 라는 생각이 들었습니다. 
또한, 마지막 쇼핑 주문 미션에서 livedata 를 어디에서도 옵저빙 하지 않을 경우, 직접 값을 꺼내쓰면 null 이 된다는 것을 확인했습니다. 이를 통해 ViewModel이 가진 상태는 반드시 옵저빙하거나, ViewModel을 통해 전달받는 방식으로 처리해야 안전하다는 점도 함께 느꼈기에 직접 값을 꺼내는 방식은 안전하지 않아 문제가 될 소지가 될 것 같습니다.

View → ViewModel(Event업데이트) → View 처리 구조로 일관성 있게 구현하면, 뷰모델은 이벤트를 업데이트하며 데이터를 넘기기만 하고, 뷰는 이벤트와 전달받은 데이터를 기반으로 화면을 업데이트할 수 있게 됩니다.

이 흐름을 따르면, ViewModel은 UI 상태 변화에 대한 트리거만 담당하고, 실제 UI 변경은 View가 관리해 MVVM 아키텍처에 맞는 책임 분리가 이루어질 것 같습니다.

이런 방식으로 리팩토링을 진행하면서, View가 ViewModel의 함수를 호출하더라도 그 함수명이 View에 대해 알지 않는 방향으로 지어져야 한다는 점도 함께 고민하게 되었습니다.

또한, 한 뷰에서 이벤트가 무수히 많아진다면, 비슷한 성격을 가진 것들끼리 이벤트를 분리해도 될 것 같단 생각이 들었습니다. 

따라서 모든 화면에서 직접 값을 꺼내쓰는 부분을 이벤트로 수정 후 ViewModel에서 일관되게 업데이트 + 값 전달 및 뷰에서 옵저빙 + 로직 처리 할 수 있도록 수정했습니다.

46b87c7d8f3d0b9f98d4fe87a38c3e8951f02631

laco-dev commented on app/src/main/java/woowacourse/shopping/view/productDetail/ProductDetailViewModel.kt:
```diff
@@ -19,13 +22,55 @@ class ProductDetailViewModel(
+    val recentWatchingProduct: LiveData<Product> get() = _recentWatchingProduct
+
+    private val _recentProductBoxVisible: MutableLiveData<Boolean> = MutableLiveData(false)
+    val recentProductBoxVisible: LiveData<Boolean> get() = _recentProductBoxVisible
```
여러개의 LiveData를 조합해야 한다면, MediatorLiveData라는 것을 학습해볼 수 있을 것 같네요.

m6z1 replied:
MediatorLiveData 를 통해서 직접적으로 값을 업데이트 시키지 않아도 알아서 다른 값을 구독해 변경될 수 있도록 수정했습니다!

9419b5290966cccee3c2ec4153cec5690281a3f6

laco-dev commented on app/src/main/java/woowacourse/shopping/view/product/ProductsActivity.kt:
```diff
@@ -10,17 +12,57 @@ import androidx.recyclerview.widget.GridLayoutManager
+                    viewModel.updateSelectedQuantity(listOf(updateQuantityProduct, recentProduct))
+                    navigateToRecentProduct(recentProduct)
+                }
+            }
```
이어서 이야기하면 모든 화면 흐름의 이동 관계를 finish() 이후 ProductsActivity가 모두 제어하는 것은 좋지 않아요.

가령 상품 화면을 거치지 않고 상세 화면으로 바로 이동할지도 모릅니다.
이 경우 해당 동작이 발생하면 상품 화면으로 돌아가지 않고 닫히니까요.
강의 자료 4단계의 힌트 Activity Flag를 참고해보세요 :)

m6z1 replied:
ProductsActivity는 결과만 받아 UI를 업데이트하도록 수정했습니다.

또한, `상품 화면 → A 상품 상세 → B 상품 상세 흐름`에서 A 상품 상세 및 B 상품 상세에서 장바구니 추가 시, 상품 화면으로 돌아갔을 때 A 상품과 B 상품 모두 상품 화면에서 수량이 반영되어야 한다는 점을 고려해,
A 상품 상세 화면에서 B 상품 상세 화면으로 이동할 때도 registerForActivityResult를 사용하도록 수정했습니다.

이를 통해,
- B 상품 상세 화면 → 소멸
- A 상품 상세 화면 → 소멸
- ProductsActivity → 결과 수신 및 UI 업데이트
의 흐름으로 돌아오며, 필요한 데이터 변경이 모두 반영될 수 있도록 개선했습니다.

75fa1e5ba7e02e7c11a8f316468a9988e59f193a

laco-dev commented on app/src/main/java/woowacourse/shopping/view/product/ProductsViewModel.kt:
```diff
@@ -5,58 +5,289 @@ import androidx.lifecycle.MutableLiveData
+                        }
+                    _products.postValue(updated)
+                    updateShoppingCartQuantity()
+                }
```
비즈니스 로직 이후 UI 상태 업데이트까지 흐름이 정리가 잘 되어있기는 한데요.
코드가 아직도 많이 복잡하다고 느껴집니다.

당장 이 함수의 depth만 봐도 그렇구요.
적절한 함수 분리로 가독성을 개선해보면 좋을 것 같아요.

m6z1 replied:
함수 네이밍 변경 및 함수 분리를 통해서 가독성을 개선했습니다!

511c5fbf3c0d706a79df1dd102785efd5b2a8bf4

laco-dev reviewed (changes requested):
3, 4단계 고생 많으셨습니다.

전반적으로 구조가 잘 정리되어 있지만 가독성이 다소 아쉬워진 것 같아요.
화면 이동에 대한 흐름도 주의할 점이 있기에 의견을 남겨두었습니다.

남겨드린 것 처럼 모든 화면 이동 Flow를 한 화면에서 제어하려고 하면 안되요.
액티비티의 Task 관리를 강의자료에서 소개한 이유이기도 합니다.

m6z1 committed:
refactor: Local DB 하나로 수정

m6z1 committed:
refactor: 최근 본 상품 박스 가시성 MediatorLiveData 로 개선

m6z1 committed:
refactor: 수량 변경 컴포넌트 데이터 값 데이터바인딩으로 수정

m6z1 committed:
test: 변경된 뷰모델 함수에 대한 테스트 수정

m6z1 committed:
chore: 테스트 패키지 프로덕션 패키지와 동일하도록 수정

m6z1 committed:
refactor: remote 용 dto 생성

m6z1 committed:
refactor: 수량 조절 버튼 공용 style 추출

m6z1 committed:
refactor: Listener 인터페이스 및 함수 네이밍 수정

m6z1 committed:
refactor: 상품 상세 ViewModel 생성자 매개변수에 product 추가

m6z1 committed:
refactor: 최근 본 상품 정보 조회 방식 개선

- 기존: 최근 본 상품 객체로 저장 및 조회
- 변경: 상품Id 를 통해 상품 정보 조회

m6z1 committed:
refactor: MockProductServer 스레드 실행 위치 변경

m6z1 committed:
chore: 불필요한 import 제거

m6z1 committed:
refactor: 장바구니 수량 출력 데이터바인딩으로 수정

m6z1 committed:
refactor: 이벤트 처리 방식 변경

- 각 화면에서 이벤트 발생 시 ViewModel에서 LiveData를 통해 Activity로 전달하도록 수정
- Activity에서는 이벤트를 관찰 및 이벤트 종류에 따른 로직 수행

m6z1 committed:
refactor: ProductsViewModel 가독성 개선

m6z1 committed:
fix: 최근 본 상품 아이템 처음 추가 될 때 상품 첫 아이템 잘리는 현상 해결

m6z1 committed:
chore: 불필요한 import 문 제거

m6z1 committed:
refactor: 화면 흐름 제어 개선

m6z1 reviewed (commented):
안녕하세요 페로로!
마지막 리뷰 요청이 정말 많이 늦어졌습니다 🥲

이번 장바구니 미션을 통해 많은 걸 깨달을 수 있었고,
페로로께서 남겨주신 정성스러운 리뷰 덕분에 많이 배우고 성장할 수 있었습니다. 정말 감사드립니다 🙇‍♀️

리뷰에서 배운 내용 3레벨 팀플 및 수료 후에도 꼭 적용하겠습니다!

다음에 또 뵐 수 있는 기회가 있다면 정말 좋겠습니다! 🥹

감사합니다

laco-dev reviewed (approved):
피듭개 반영하느라 고생 많으셨습니다.
도움이 되셨다니 다행입니다. 작성해주신 것 처럼 지금까지 배운 내용들을 잘 챙겨두시면 앞으로 많이 도움이 될 것 같아요.

앞으로도 응원하겠습니다.
